feat: Inline Git Blame (#13133)
helix-editor/helix: book/src/editor.md

[^2]: You may also have to activate them in the language server config for them to appear, not just in Helix. Inlay hints in Helix are still being improved on and may be a little bit laggy/janky under some circumstances. Please report any bugs you see so we can fix them!

### `[editor.inline-blame]` Section

Inline blame is virtual text that appears at the end of a line, displaying information about the most recent commit that affected this line.

| Key     | Description                                | Default |
| ------- | ------------------------------------------ | ------- |

> | `show` | Choose when to show inline blame | `"never"` |

Reviewer: ```suggestion
| `show` | When to show inline blame | `"never"` |
```

> | `auto-fetch` | Choose when inline blame should be computed | `false` |

Reviewer: ```suggestion
| `auto-fetch` | Automatically fetch blame information in the background | `false` |
```

> | `format` | The format in which to show the inline blame | `"{author}, {time-ago} • {title} • {commit}"` |

Reviewer: ```suggestion
| `format` | Inline blame message format | `"{author}, {time-ago} • {title} • {commit}"` |
```

> `show` can be one of the following:
> - `"all-lines"`: Inline blame is on every line.
> - `"cursor-line"`: Inline blame is only on the line of the primary cursor.
> - `"hidden"`: Inline blame is hidden.

Reviewer: ```suggestion
`show` can be one of the following:
- `"always"`: Inline blame is on every line.
- `"cursor-line"`: Inline blame is only on the line of the primary cursor.
- `"never"`: Inline blame is never shown inline.
```

Author: I don't think `"always"` is descriptive enough to mean "show on every line". The `"cursor-line"` option also "always" shows the inline blame but only on a single line

> 
> Inline blame will only show if the blame for the file has already been fetched.

Reviewer: ```suggestion
```

> The `auto-fetch` key determines under which circumstances the blame is fetched, and can be one of the following:
> - `false`: Blame for the file is fetched only when explicitly requested, such as when using `space + B` to blame the line of the cursor. There may be a little delay when loading the blame.
> 
>   When opening new files, even with `show` set to `"all-lines"` or `"cursor-line"`, the inline blame won't show. It needs to be fetched first in order to become available, which can be triggered manually with `space + B`.
> - `true`: Blame for the file is fetched in the background.
> 
>   This will have zero effect on performance of the Editor, but will use a little bit extra resources.
> 
>   Directly requesting the blame with `space + B` will be instant. Inline blame will show as soon as the blame is available when loading new files.

Reviewer: ```suggestion
By default (or with `auto-fetch` set to false), blame for the current file is fetched only when explicitly requested, such as when using `space + B` to display the blame for the line of the cursor. There may be a little delay when loading the blame. When `auto-fetch` is enabled, blame for the file is fetched in the background; this will have zero effect on performance of the editor, but will use a little bit extra resources. Directly requesting the blame with `space + B` will be instant. Inline blame will show as soon as the blame is available when loading new files.

When opening new files, even with `show` set to `"all-lines"` or `"cursor-line"`, the inline blame won't show. It needs to be fetched first in order to become available, which can be triggered manually with `space + B`.
```

Reviewer: > When opening new files, even with `show` set to `"all-lines"` or `"cursor-line"`, the inline blame won't show. It needs to be fetched first in order to become available, which can be triggered manually with `space + B`.

This last sentence here feels like something ideally we could change the structure of our configuration for to avoid, where the `show` setting doesn't *not* work when the other option is disabled.

[^2]: You may also have to activate them in the language server config for them to appear, not just in Helix. Inlay hints in Helix are still being improved on and may be a little bit laggy/janky under some circumstances. Please report any bugs you see so we can fix them!

### `[editor.inline-blame]` Section

Inline blame is virtual text that appears at the end of a line, displaying information about the most recent commit that affected this line.

| Key     | Description                                | Default |
| ------- | ------------------------------------------ | ------- |

> | `behaviour` | Choose when to show inline blame | `"hidden"` |

Reviewer: ```suggestion
| `behavior` | Choose when to show inline blame | `"hidden"` |
```
Should we use American English for configuration options?

| `compute` | Choose when inline blame should be computed | `"on-demand"` |

> | `format` | The format in which to show the inline blame | `"{author}, {time-ago} • {message} • {commit}"` |

Reviewer: Due to the complexity of this format feature it might make sense to extract this into a follow-up pull request? Lightening up this pull request could also ease reviewing and give it a better chance of merging expediently :)

Author: I wouldn't say it's complex,it only adds like 80 lines of code. 

There are some bits of information about a commit some might want to know but doesn't make sense to include for everyone, such as email of the author

I can remove or put it in a different PR if really necessary, but for now I'll leave it since it's already implemented

Reviewer: > I wouldn't say it's complex,it only adds like 80 lines of code.

Well it's complex in that it introduces more surface area for review. People will have different opinions on the names of the available variables, might want more/less exposed variables, the behavior needs better documentation and might need to be reworked, etc.

The `behaviour` can be one of the following:
- `"all-lines"`: Inline blame is on every line.
- `"cursor-line"`: Inline blame is only on the line of the primary cursor.

> - `"hidden"`: Inline blame is not shown.

Reviewer: I would prefer an option named something more like `show`/`render`, with available values `all`/`always`, `cursor`/`cursor-line`, and `none`/`never`/`disabled`.

```toml
[editor.inline-blame]
show = "always"
```

```toml
[editor.inline-blame]
render = "cursor-line"
```

```toml
[editor.inline-blame]
render = "never"
```

> Inline blame will only show if the blame for the file has already been computed.

Reviewer: This line seems disconnected and I'm not sure what it refers to? Is this for a specific `behavior` option?

The `compute` key determines under which circumstances the blame is computed, and can be one of the following:
- `"on-demand"`: Blame for the file is computed only when explicitly requested, such as when using `space + B` to blame the line of the cursor. There may be a little delay when loading the blame. When opening new files, even with `behaviour` not set to `"hidden"`, the inline blame won't show. It needs to be computed first in order to become available. This computation can be manually triggered by requesting it with `space + B`.

> - `"background"`: Blame for the file is loaded in the background. This will have zero effect on performance of the Editor, but will use a little bit extra resources. Directly requesting the blame with `space + B` will be instant. Inline blame will show as soon as the blame is available when loading new files.

Reviewer: It would make more sense to me to have the default behavior of `editor.git-blame` be shown "on demand" and have a boolean configuration option along the lines of `editor.inline-blame.background` or `editor.inline-blame.auto-fetch` to enable the "background" feature.

Author: So when you press space + b to show line for the current file, then you have see inline blame for the cursor line?

I think that will be unexpected, I don't assume everyone that press Space B has read the docs about inline blame

Whereas in the reverse, you explicitly turn it on so there are no surprises

Reviewer: > So when you press space + b to show line for the current file, then you have see inline blame for the cursor line?

No, where does I say that?  When you press space + b with `editor.inline-blame.auto-fetch` not enabled, it will have a delay as it fetches the blame data on demand. When you press space + b with `editor.inline-blame.auto-fetch` *enabled*, it will show instantly because the blame data has already been fetched in the background. Whether or not it displays in the cursor line is controlled by the other setting, `behavior` as it is called now.

> `inline-blame-format` allows customization of the blame message, and can be set to any string. Variables can be used like so: `{variable}`. These are the available variables:

Reviewer: ```suggestion
Set a format string for `format` to customize the blame message displayed. Variables are text placeholders wrapped in curly braces (`{variable}`). The following variables are available:
```

- `author`: The author of the commit
- `date`: When the commit was made
- `time-ago`: How long ago the commit was made

> - `message`: The message of the commit, excluding the body

Reviewer: ```suggestion
- `title`: The title of the commit message (excluding the body)
```

- `body`: The body of the commit
- `commit`: The short hex SHA1 hash of the commit
- `email`: The email of the author of the commit

> Any of the variables can potentially be empty.

Reviewer: Why would `author` be empty? Why would date or time-ago be empty? What about `commit`? It seems that only `message` (or `title`, as I suggested), `body`, and `email` could be empty?

Author: I agree it's quite confusing. I've simply removed mention of this behaviour. 99% of people reading it may get confused even though this isn't something they should worry about

Anything can be empty because all of the functions for obtaining info like the `Commit::author` are fallible (return `Result<...>`)

In this case, the content before the variable will not be included in the string.

> If the variable is at the beginning of the string, the content after the variable will not be included.

Reviewer: This behavior seems quite confusing and is hard to interpret.

EDIT: It's also unclear what "content" means in this context, on second thought. I didn't realize until a few minutes later but I guess you mean the non-variable text up until the next variable? In the first example of `author` being empty, the content after the `{author}` interpolated variable but before the subsequent `{time-ago}` variable is what is removed.

Author: I agree the wording here could be improved
